lib/modules/apostrophe-doc-type-manager/lib/api.js

```diff
@@ -242,13 +242,23 @@ module.exports = function(self, options) {
   // Return a new schema containing only fields for which the
   // current user has the permission specified by the `permission`
   // property of the schema field, or there is no `permission` property for the field.
+  // property permission and readOnly also accept a callback
 
   self.allowedSchema = function(req) {
     var disabled;
     var type;
     var schema = _.filter(self.schema, function(field) {
+      if (_.isFunction(field.permission)) {
+        return !!(field.permission(self, req, field));
+      }
       return (!field.permission) || self.apos.permissions.can(req, field.permission);
     });
+    schema = schema.map(function(field) {
+      if (_.isFunction(field.readOnly)) {
+        field.readOnly = !!(field.readOnly(self, req, field));
```

Comment:
You are modifying `field` so everyone will get the same answer in the future, which is not what you want. To be safe you would need to make a shallow clone of `field` before modifying it.